src/inet/UDPEndPointImplOT.cpp

```diff
@@ -0,0 +1,264 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    Copyright (c) 2018 Google LLC.
+ *    Copyright (c) 2013-2018 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * This file implements Inet::UDPEndPoint using OpenThread.
+ */
+
+#include <inet/UDPEndPointImplOT.h>
+
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <platform/OpenThread/OpenThreadUtils.h>
+//#include <include/platform/ThreadStackManager.h>
+
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace Inet {
+
+void UDPEndPointImplOT::handleUdpReceive(void * aContext, otMessage * aMessage, const otMessageInfo * aMessageInfo)
+{
+    UDPEndPointImplOT * ep = static_cast<UDPEndPointImplOT *>(aContext);
+    IPPacketInfo pktInfo;
+    static uint16_t msgReceivedCount = 0;
+    uint16_t msgLen                  = otMessageGetLength(aMessage);
+    System::PacketBufferHandle payload;
+    char sourceStr[Inet::IPAddress::kMaxStringLength];
+    char destStr[Inet::IPAddress::kMaxStringLength];
+
+    if (msgLen > System::PacketBuffer::kMaxSizeWithoutReserve)
+    {
+        return;
+    }
+
+    payload = System::PacketBufferHandle::New(msgLen, 0);
+
+    pktInfo.SrcAddress  = chip::DeviceLayer::Internal::ToIPAddress(aMessageInfo->mPeerAddr);
+    pktInfo.DestAddress = chip::DeviceLayer::Internal::ToIPAddress(aMessageInfo->mSockAddr);
+    pktInfo.SrcPort     = aMessageInfo->mPeerPort;
+    pktInfo.DestPort    = aMessageInfo->mSockPort;
+
+    if (payload.IsNull())
+    {
+        return;
+    }
+    pktInfo.SrcAddress.ToString(sourceStr, Inet::IPAddress::kMaxStringLength);
+    pktInfo.DestAddress.ToString(destStr, Inet::IPAddress::kMaxStringLength);
+
+    ChipLogDetail(Inet,
+                  "UDP Message Received packet nb : %d with the following data :\r\nSrcAddr : %s\r\nSrc Port : %d\r\n\r\nDestAddr "
+                  ": %s\r\nDest Port %d\r\nPayload Length %d",
+                  ++msgReceivedCount, sourceStr, pktInfo.SrcPort, destStr, pktInfo.DestPort, msgLen);
+
+    memcpy(payload->Start(), &pktInfo, sizeof(IPPacketInfo));
+
+    if (otMessageRead(aMessage, 0, payload->Start() + sizeof(IPPacketInfo), msgLen) != msgLen)
+    {
+        return;
+    }
+    payload->SetDataLength(msgLen + sizeof(IPPacketInfo));
+
+    ep->Retain();
```

Comment:
I checked the `Retain` implementation in `src/lib/core/ReferenceCounted.h`:

``` C++
    /** Adds one to the usage count of this class */
    Subclass * Retain()
    {
        if (mRefCount == std::numeric_limits<count_type>::max())
        {
            abort();
        }
        ++mRefCount;

        return static_cast<Subclass *>(this);
    }
```

Unfortunately it uses `++mRefCount` rather than atomic operations. Since this function is typically called in the OpenThread task, this will case potential race conditions. Please file an issue to replace the implementation with an atomic one.

Reply:
issue #13300 created

Reply:
`ReferenceCounted` class is designed to be used to wrapping CHIP object, which is only mutable inside chip thread. 

There used to be a `AtomicReferenceCounted` class, which somehow being removed, may be we can bring it back.

Reply:
It's weird that this conversation is still happening, since this section is a simple copy-paste from the `UdpEndpointImplLwIP`. If this is really a huge problem, why wasn't it fixed in the other implementation ? and if it's not a problem in the LwIP implementation what's missing in this one to fix it ? @kghost @gjc13 ?

Reply:
The situation is different from `UdpEndpointImplLwIP`, because in that implementation, the callback is called inside the chip mainloop, but here, I'm not sure whether the callback is called inside chip mainloop.

Reply:
From the trace I did with UdpEndpointImplLwIP, upon reception of an IPV6 packet from the Thread task, it is forwarded to the LwIP task and the callback is executed from the LwIP task and not from the mainloop, but maybe I missed something.

Reply:
I need to recheck it. If in this case, both of the implementation is problematic, we need to find a way to work it out. Using a atomic reference counter is not going to help here.

Reply:
If it's the same problem in both implementation, I suggest that we fix it in another PR, since it's now a known issue that was present for a long time. I wouldn't want the Multicast feature to be in jeopardy because of a past error that was never fixed.

Reply:
closing with issue #13458 opened